libcloud/common/base.py

```diff
@@ -46,11 +48,48 @@
 
 from libcloud.utils.misc import lowercase_keys
 from libcloud.utils.compression import decompress_data
+from libcloud.common.exception import RateLimit, exception_from_message
 from libcloud.common.types import LibcloudError, MalformedResponseError
-
 from libcloud.httplib_ssl import LibcloudHTTPConnection
 from libcloud.httplib_ssl import LibcloudHTTPSConnection
 
+DEFAULT_TIMEOUT = 3
+DEFAULT_SLEEP = 3
+DEFAULT_BACKCOFF = 1
+
+
+def retry(*dargs, **dkwargs):
+    """
+    Retry decorator helps to handle common exception.
+
+    Usage::
+            @retry
+            @retry(timeout=30)
+            @retry(timeout=10, sleep=5)
+    """
+    def decorator(func):
+        def wrapped(*args, **kwargs):
+            attempts = 0
+            timeout = dkwargs.get('timeout', DEFAULT_TIMEOUT)
+            bsleep = dkwargs.get('sleep', DEFAULT_SLEEP)
+            backoff = dkwargs.get('backoff', DEFAULT_BACKCOFF)
+            end = datetime.now() + timedelta(seconds=timeout)
+            while datetime.now() < end:
+                try:
+                    return func(*args, **kwargs)
+                except (socket.error,  socket.gaierror, httplib.NotConnected,
+                        httplib.ImproperConnectionState, RateLimit) as exc:
+                    if (isinstance(exc, RateLimit) and
+                            exc.retry_after is not None):
+                        time.sleep(exc.retry_after)
```

Comment:
This will most likely trigger the timeout.  We might need to also update ```end``` based on this value:

```end = datetime.now() + timedelta(seconds=exc.retry_after + timeout)```